### PR TITLE
Fix #12159 (Numeral Notations do not play well with multiple scopes for the same inductive)

### LINKED ISSUE #12159: Numeral Notations do not play well with multiple scopes for the same inductive
#### Description of the problem
```coq
Declare Scope A.
Declare Scope B.
Delimit Scope A with A.
Delimit Scope B with B.
Definition to_unit (v : Decimal.uint) : option unit
  := match Nat.of_uint v with O => Some tt | _ => None end.
Definition of_unit (v : unit) : Decimal.uint := Nat.to_uint 0.
Numeral Notation unit to_unit of_unit : A.
Numeral Notation unit to_unit of_unit : B.
Check tt. (* 0%B : unit *)
Open Scope A.
Check tt. (* 0%B : unit *) (* seems wrong *)
Undelimit Scope B.
Check tt. (* tt : unit *) (* seems wrong *)
```

cc @herbelin @proux01 

#### Coq Version

8.11.1

### PR DESCRIPTION
**Pull request by proux01**
**Kind:** bug fix

Fixes #12159 

- [x] Added / updated test-suite
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Review by herbelin — COMMENTED** [review]
Thanks a lot for addressing this. I suspect that some fine-tuning remains to be done in selecting the printing rule:
- the `LonelyNotationItem` case should not be discarded I think;
- the `InnerPrimToken.do_uninterp` may fail for the given term and it may be worth to then try another scope, i.e. to interleave `InnerPrimToken.do_uninterp` with the search in the `scope_stack` (of course, that's also a question of how to resolve ambiguities in a most natural or expected way).

From a longer term perspective, note also the following issues:
- one may consider to eventually interleave primitive token notations and ordinary notations 
- the scope order strategy is not always what is wanted for printing; see the long discussion at #8187 and #9207 which I'm unfortunately not able to summarize; citing it only to tell that there are open questions

**Comment by proux01** [comment]
Thanks to both of you for your quick and precise feedback. Here is a new version taking them into account. It works as follows:
1. keep only uninterpreters that do not fail
1. then keep only uninterpretations that have a delimiter (as noted by Jason)
1. select according to `scope_stack`, otherwise take the first one (the last defined notation)

**Comment by JasonGross** [comment]
> 2\. then keep only uninterpretations that have a delimiter (as noted by Jason)

I think this is not quite right.  If there is no delimiter but the scope is open, the notation should still be able to be printed.  (And we should make sure we have a test for this case.)

File: doc/changelog/03-notations/12163-fix-12159.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Fixed:**
+  Fix #12159 (Numeral Notations do not play well with multiple scopes
+  for the same inductive)
+  (`#12163 <https://github.com/coq/coq/pull/12163>`_,
+  by Pierre Roux).
```
Comment (JasonGross) [inline doc/changelog/03-notations/12163-fix-12159.rst:5]: I think it's standard to link to the bug being fixed and mention it in the parenthetical
```suggestion
  Numeral Notations now play better with multiple scopes for the same
  inductive ( `#12163 <https://github.com/coq/coq/pull/12163>`_, fixes
  `#12159 <https://github.com/coq/coq/pull/12159>`_, by Pierre Roux).
```

**Review by JasonGross — CHANGES_REQUESTED** [review]

**Comment by JasonGross** [comment]
> > 2. then keep only uninterpretations that have a delimiter (as noted by Jason)
> 
> I think this is not quite right. If there is no delimiter but the scope is open, the notation should still be able to be printed. (And we should make sure we have a test for this case.)

Ah, but it looks like this is already the behavior that shows up in the test-suite, so perhaps just the description is wrong?  Also, should we say what the new behavior is in the changelog?

**Comment by JasonGross** [comment]
Though I guess it's also not clear to me why numeral notations follow a different codepath than the rest of the notations.  It seems like perhaps the codepaths should be unified?  @herbelin, can you comment on this?

(This is likely out of scope for this PR, though, which I'm in favor of having merged.)

**Comment by proux01** [comment]
> the description is wrong?

You're right, it should read "have a delimiter if needed"

> Also, should we say what the new behavior is in the changelog?

Right, I'm updating it.

> Though I guess it's also not clear to me why numeral notations follow a different codepath than the rest of the notations. It seems like perhaps the codepaths should be unified?

Indeed maybe. I'm unfortunately not familiar enough with this code to come up with a nice unification.

File: doc/changelog/03-notations/12163-fix-12159.rst [hunk 2]
```diff
@@ -0,0 +1,11 @@
+- **Fixed:**
+  Numeral Notations now play better with multiple scopes for the same
+  inductive. Previously, when multiple numeral notations where defined
+  for the same inductive, only the last one was considered for
+  printing. Now, among the printers that produce an output and have a
+  scope delimiter (if required), the selection is made according to
+  open scopes (or the last defined one if no appropriate scope is
+  open)
```
Comment (JasonGross) [inline doc/changelog/03-notations/12163-fix-12159.rst:8]: I think this is a bit clearer and more explicit:
```suggestion
  printing. Now, among the printers that produce an output and either
  have a scope delimiter or are open, the selection is made according
  to the order of opened scopes, or according to the last defined
  scope if no appropriate scope is open
```

**Review by JasonGross — APPROVED** [review]
Changelog and test-suite behavior look good to me (feel free to take or leave my changelog suggestion)

File: doc/changelog/03-notations/12163-fix-12159.rst (hunk 2 above)
Comment (proux01) [reply doc/changelog/03-notations/12163-fix-12159.rst]: Thanks, done (just replaced scopes by notation for the last default choice).

**Comment by proux01** [comment]
I forgot to take local scopes (argument scopes for instance) into account. Just fixed that and added a test.

**Comment by Zimmi48** [comment]
@herbelin and @JasonGross: one of you should take this PR as assignee.

**Comment by JasonGross** [comment]
I would prefer that @herbelin take it, as I'm not very knowledgeable about the part of the code-base this touches, just what the behavior should be.

**Review by herbelin — APPROVED** [review]
This seems ok to me. (By which I mean that the rules used to prioritarily select which notation to use for printing seems reasonable.)

If you have the energy, a test for lonely notations may be worth too.

Unless @JasonGross (or someone else) has more to say, I'll merge on Saturday.

**Comment by proux01** [comment]
> a test for lonely notations may be worth too.

I tried this
```Coq
Declare Scope A.
Delimit Scope A with A.
Definition to_unit (v : Decimal.uint) : option unit
  := match Nat.of_uint v with 1 => Some tt | _ => None end.
Definition of_unit (v : unit) : Decimal.uint := Nat.to_uint 1.
Numeral Notation unit to_unit of_unit : A.

Open Scope A.  (* scope A *)
Check tt.
Notation "1" := true.  (* lonely notation put on top of A *)
Check tt.
```
and ended up with `1 : unit`. Do we agree that this should be `1%A` so that it can be correctly reparsed? In which case I should fix it.

**Comment by herbelin** [comment]
> Do we agree that this should be 1%A

I agree.

**Comment by proux01** [comment]
Fixed and test added.

Remark: I'm not very proud of the fix but as noted by Jason above, this code would maybe need some later refactoring to unify it with regular notations.

**Comment by herbelin** [comment]
I'm not sure looking at masked lonely notations only is enough. One can have a notation `1` in scope `A` for something else than `tt` and then have to take also into account that `1` in scope `B` has to be delimited when scope `A` is opened more recently than `B`.

But I'm also ok in considering this code as an intermediate state before a merge of printing of primitive notations and printing of ordinary notations.

**Comment by proux01** [comment]
> One can have a notation `1` in scope `A` for something else than `tt` and then have to take also into account that `1` in scope `B` has to be delimited when scope `A` is opened more recently than `B`.

I agree. This should probably be properly fixed when merging with ordinary notations rather than adding yet another hack. Meantime I just put a comment and an (incorrect) test in order not to forget that point.

> But I'm also ok in considering this code as an intermediate state before a merge of printing of primitive notations and printing of ordinary notations.

I may give it a try at some point (no idea when though) but that would probably come with a number of questions and would definitely need a careful review.

**Comment by proux01** [comment]
CI green

**Comment by herbelin** [comment]
So, even if it does not solve all issues with multiple scopes, it improves and I'm leaning to merge the PR. If no further objections, I'll merge on Friday.

**Pull request merged**